runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/GroupIntoBatchesOverride.java

```diff
@@ -87,14 +94,25 @@ private BatchGroupIntoBatches(long batchSize) {
                   new DoFn<KV<K, Iterable<V>>, KV<K, Iterable<V>>>() {
                     @ProcessElement
                     public void process(ProcessContext c) {
-                      // Iterators.partition lazily creates the partitions as they are accessed
-                      // allowing it to partition very large iterators.
-                      Iterator<List<V>> iterator =
-                          Iterators.partition(c.element().getValue().iterator(), (int) batchSize);
-
-                      // Note that GroupIntoBatches only outputs when the batch is non-empty.
-                      while (iterator.hasNext()) {
-                        c.output(KV.of(c.element().getKey(), iterator.next()));
+                      List<V> currentBatch = Lists.newArrayList();
+                      long batchSizeBytes = 0;
+                      for (V element : c.element().getValue()) {
+                        currentBatch.add(element);
+                        if (weigher != null) {
+                          batchSizeBytes += weigher.apply(element);
+                        }
+                        if (currentBatch.size() == maxBatchSizeElements
+                            || (maxBatchSizeBytes != Long.MAX_VALUE
+                                && batchSizeBytes >= maxBatchSizeBytes)) {
+                          c.output(KV.of(c.element().getKey(), currentBatch));
+                          // Call clear() since that allows us to reuse the array memory for
+                          // subsequent batches.
+                          currentBatch.clear();
```

Comment:
looking at this, I don't understand how this is safe.  Isn't this mutating an element once it's been emitted? If fused step happens to store this list somewhere outside of the processElement call (state, buffer between start/end bundle, etc), it'll be mutated out from under it.

Is there a copy happening somewhere else under the covers that makes this safe?

Reply:
oof, this is a bug.

It's probably ok in many cases, since the fused fn will be executed immediately when output is called. However the fused fn is allowed to store the element in memory and process it in finishBundle, and if that happens the data will be stomped. We need to remove this buggy optimization :P